### ISSUE TITLE
Fail to build with Clang on Linux

### ISSUE THREAD

**Issue by petterreinholdtsen** [CasparCG/server#1141]
I tried building the current git code with clang on Linux, but there are several code errors blocking this from working.  Perhaps the automatic code test should be extended to also build using clang to verify the code is compatible with both gcc and clang?

Here is an example of the changes needed to build log.cpp with clang:

```
diff --git a/src/common/log.cpp b/src/common/log.cpp
index 2c188e8..bdb111b 100644
--- a/src/common/log.cpp
+++ b/src/common/log.cpp
@@ -23,6 +23,9 @@
 #include "except.h"
 
 #include <boost/algorithm/string/predicate.hpp>
+#include <boost/algorithm/string/replace.hpp>
+#include <boost/core/null_deleter.hpp>
+#include <boost/filesystem/operations.hpp>
 #include <boost/locale.hpp>
 #include <boost/log/attributes/attribute_value.hpp>
 #include <boost/log/attributes/function.hpp>
```

Without this change, boost::algorithm::replace_all_copy() and other symbols are not found by the compiler.

**Comment by petterreinholdtsen** [comment]
There are two build errors left when using clang.  The first is this:

```
/home/user/src/casparcg-server-git/src/accelerator/ogl/util/device.cpp:199:31: error: no
      member named 'emplace' in
      'tbb::concurrent_bounded_queue<std::shared_ptr<caspar::accelerator::ogl::buffer>,
      tbb::cache_aligned_allocator<std::shared_ptr<caspar::accelerator::ogl::buffer> > >'
            self->sync_queue_.emplace(std::move(buf));
            ~~~~~~~~~~~~~~~~~ ^
1 warning and 1 error generated.
```

I do not know how to fix it.  The emplace() method is only conditionally included in <tbb/concurrent_queue.h>, and I do not know the correct way to activate it.

The second problem (when commenting out the problematic call above), is linking failure, like this:

```
[100%] Linking CXX executable casparcg
/usr/bin/ld: /usr/bin/../lib/gcc/x86_64-linux-gnu/8/../../../x86_64-linux-gnu/libboost_locale.a(numeric.o): undefined reference to symbol '_ZNSt5ctypeIwED2Ev@@GLIBCXX_3.4'
/usr/bin/ld: //usr/lib/x86_64-linux-gnu/libstdc++.so.6: error adding symbols: DSO missing from command line
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

I have no idea how to fix it.

I recommend to reopen this issue until the build succeed.

**Comment by petterreinholdtsen** [comment]
I managed to figure out the linker problem.  The cause was the use of clang instead of clang++ when calling cmake.  When calling it like this, the linker problem go away:

```
CC=clang CXX=clang++ cmake -DENABLE_HTML=OFF ../src/ && make
```

The other problem with emplace() still exist, and I worked around it by simply commenting out the problematic call:

```
--- a/src/accelerator/ogl/util/device.cpp
+++ b/src/accelerator/ogl/util/device.cpp
@@ -196,7 +196,7 @@ struct device::impl : public std::enable_shared_from_this<impl>
 
         auto ptr = buf.get();
         return std::shared_ptr<buffer>(ptr, [buf = std::move(buf), self = shared_from_this()](buffer*) mutable {
-            self->sync_queue_.emplace(std::move(buf));
+            //self->sync_queue_.emplace(std::move(buf));
         });
     }
 
```

Last, there is a code error where 'int*' is used where 'bool*' is expected.  I solved it like this:

```
--- a/src/modules/decklink/producer/decklink_producer.cpp
+++ b/src/modules/decklink/producer/decklink_producer.cpp
@@ -378,7 +378,7 @@ class decklink_producer : public IDeckLinkInputCallback
         graph_->set_text(print());
         diagnostics::register_graph(graph_);
 
-        int status = 0;
+        bool status = 0;
         int flags  = bmdVideoInputEnableFormatDetection;
 
         if (!format.empty()) {
```

With these three parts in place, casparcg builds with clang.  There are a few compiler warnings to look at (typically unused enumeration values), but none of them seem fatal.

**Comment by Julusian** [comment]
> I managed to figure out the linker problem.

Thats good, that linker issue looked like a horrible one to figure out.

> Last, there is a code error where 'int*' is used where 'bool*' is expected

This might cause a build error on windows now, as I think the api uses ints there (I have seen that in other places in blackmagic apis)

The compiler warnings will be nice, most likely things we want to fix anyway :)

**Comment by petterreinholdtsen** [comment]
OK.  Here are the warnings:

```
/home/frikanalen/src/casparcg-server-git/src/core/diagnostics/osd_graph.cpp:153:21: warning: 
      18 enumeration values not handled in switch: 'LostFocus', 'GainedFocus',
      'TextEntered'... [-Wswitch]
            switch (e.type) {
                    ^
/home/frikanalen/src/casparcg-server-git/src/core/producer/frame_producer.cpp:89:20: warning: 
      'last_frame' overrides a member function but is not marked 'override'
      [-Winconsistent-missing-override]
        draw_frame last_frame() { return draw_frame{}; }
                   ^
/home/frikanalen/src/casparcg-server-git/src/core/../core/producer/frame_producer.h:101:26: note: 
      overridden virtual function is here
    virtual draw_frame   last_frame()
                         ^
/home/frikanalen/src/casparcg-server-git/src/core/producer/frame_producer.cpp:90:20: warning: 
      'first_frame' overrides a member function but is not marked 'override'
      [-Winconsistent-missing-override]
        draw_frame first_frame() { return draw_frame{}; }
                   ^
/home/frikanalen/src/casparcg-server-git/src/core/../core/producer/frame_producer.h:108:24: note: 
      overridden virtual function is here
    virtual draw_frame first_frame()
                       ^
/home/frikanalen/src/casparcg-server-git/src/modules/ffmpeg/util/av_util.cpp:176:13: warning: 
      enumeration values 'luma', 'count', and 'invalid' not handled in switch
      [-Wswitch]
    switch (format) {
            ^

```

**Comment by petterreinholdtsen** [comment]
I checked GetFlag() on Windows, and it uses BOOL as its argument type.  Perhaps this change is better?

```
diff --git a/src/modules/decklink/producer/decklink_producer.cpp b/src/modules/decklink/producer/decklink_producer.cpp
index 2c7fd2be3..c4859629a 100644
--- a/src/modules/decklink/producer/decklink_producer.cpp
+++ b/src/modules/decklink/producer/decklink_producer.cpp
@@ -378,7 +378,7 @@ class decklink_producer : public IDeckLinkInputCallback
         graph_->set_text(print());
         diagnostics::register_graph(graph_);
 
-        int status = 0;
+        BOOL status = 0;
         int flags  = bmdVideoInputEnableFormatDetection;
 
         if (!format.empty()) {
```

It build with clang.

**Comment by petterreinholdtsen** [comment]
A fix for the status problem is proposed in #1161.

**Comment by Julusian** [comment]
The emplace issue was caused by tbb not being able to figure out c++11 feature support (https://github.com/01org/tbb/issues/22). With some help, that has been resolved.

All the other warnings are now fixed too
